servicetalk-grpc-utils/src/test/java/io/servicetalk/grpc/utils/LoggingGrpcLifecycleObserverTest.java

```diff
@@ -0,0 +1,159 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.servicetalk.grpc.utils;
+
+import io.servicetalk.grpc.api.GrpcLifecycleObserver;
+import io.servicetalk.http.api.HttpHeaders;
+import io.servicetalk.http.api.HttpProtocolVersion;
+import io.servicetalk.http.api.HttpRequestMetaData;
+import io.servicetalk.http.api.HttpRequestMethod;
+import io.servicetalk.http.api.HttpResponseMetaData;
+import io.servicetalk.http.api.HttpResponseStatus;
+import io.servicetalk.logging.api.LogLevel;
+import io.servicetalk.logging.slf4j.internal.FixedLevelLogger;
+import io.servicetalk.transport.api.ConnectionInfo;
+
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.isA;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.anyLong;
+import static org.mockito.ArgumentMatchers.anyString;
+import static org.mockito.ArgumentMatchers.eq;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+class LoggingGrpcLifecycleObserverTest {
+
+    private HttpRequestMetaData mockRequestMetadata;
+    private HttpResponseMetaData mockResponseMetadata;
+    private FixedLevelLogger mockLogger;
+    private GrpcLifecycleObserver.GrpcExchangeObserver observer;
+
+    @BeforeEach
+    public void setUp() {
+        mockRequestMetadata = mock(HttpRequestMetaData.class);
+        when(mockRequestMetadata.headers()).thenReturn(mock(HttpHeaders.class));
+        when(mockRequestMetadata.method()).thenReturn(HttpRequestMethod.GET);
+        when(mockRequestMetadata.requestTarget()).thenReturn("/foo/bar");
+        when(mockRequestMetadata.version()).thenReturn(HttpProtocolVersion.HTTP_2_0);
+
+        mockResponseMetadata = mock(HttpResponseMetaData.class);
+        when(mockResponseMetadata.status()).thenReturn(HttpResponseStatus.OK);
+        when(mockResponseMetadata.headers()).thenReturn(mock(HttpHeaders.class));
+
+        mockLogger = mock(FixedLevelLogger.class);
+        when(mockLogger.logLevel()).thenReturn(LogLevel.INFO);
+
+        observer = new LoggingGrpcLifecycleObserver(mockLogger).onNewExchange();
+    }
+
+    @Test
+    void testFactoryMethod() {
+        assertThat(observer, isA(GrpcLifecycleObserver.GrpcRequestObserver.class));
+        assertThat(observer, isA(GrpcLifecycleObserver.GrpcResponseObserver.class));
+    }
+
+    @Test
+    void testOnRequestError() {
+        observer.onConnectionSelected(mock(ConnectionInfo.class));
+        Throwable requestError = new RuntimeException("fake request error");
```

Comment:
For purposes like this one we have `DeliberateException`. You will need to add
```
testImplementation testFixtures(project(":servicetalk-concurrent-internal"))
```
dependency to access it.